### DISCUSSION TITLE
why not use linked list for stack c#

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#65295] [General] [answered]
according to https://source.dot.net/#System.Collections.NonGeneric/System/Collections/Stack.cs,6acda10c5f8b128e
it's implemented by using array

**Comment by Clockwork-Muse** [comment, marked as answer]
Trivial (possibly not the reason) answer - it takes less memory.  For classes, a linked list takes at least twice the memory of an array (An array will contain one pointer to the data on the heap, but a linked list will need one pointer to the data, plus a pointer to the next node).

Things are slightly better if the contents are structs, because the contents of the struct also occupy container memory (and so don't have pointer to the data), but it still takes some extra memory.

Too, while it's true that a linked list makes it cheaper to insert/remove elements from the middle of the list, in the case of a stack this isn't necessarily helpful, because most modifications are going to happen at one end.  Note too that traditional "instruction stack space" is usually laid out in sequential address space (ie, as if it were an array).

**Reply by Xyncgas** [reply]
but adding stuffs in array might reallocate (O(N)), while adding to linked lists is always O(1)

**Reply by danmoseley** [reply]
@Xyncgas this resize cost is amortized by enlarging in multiples, eg 2x.

**Reply by John0King** [reply]
I think it may better have a struct like `ReadonlySequence<T>`  aka.  an `LinkedList<T[]>`

**Reply by Joe4evr** [reply]
You mean like [the one that already exists](https://apisof.net/catalog/b22377f2-dfbc-985f-6163-4d2ba1ad26a6)?

**Reply by John0King** [reply]
yes,  but may not exact type ，
